### DISCUSSION TITLE
Non-uniform array data structure

### DISCUSSION THREAD

**Discussion by termi-official** [Ferrite-FEM/Ferrite.jl#827] [Ideas]
I would like to quickly address the idea of a new data structure for Ferrite, which is something like `OuterNonuniformArray`. The purpose would be to have an abstraction to use continuous memory over entities which might need different amounts of data. Concretely this would be introduced for

* per-face data as in BCValues and FaceValues
* DofHandler cell dofs
* L2 projector quadrature point passing
* internal variable storage in examples

The design might be something like

```julia
struct OuterNonuniformArray{T, DT <: AbstractArray{T}, OT <: AbstractArray{<:Int}} <: AbstractArray{T}
   data::DT 
   offsets::OT
end
```

**Comment by lijas** [comment]
We did have this in the mixed dofhandler for a while: https://github.com/Ferrite-FEM/Ferrite.jl/blob/2cd15f77f4431ea1df34c9510a640636cedfd58d/src/Dofs/MixedDofHandler.jl#L51
(See the definition of CellVector above the MixedDofHandler in the link)

**Comment by termi-official** [comment]
https://github.com/JuliaArrays/ArraysOfArrays.jl might be an interesting candidate

**Comment by KristofferC** [comment]
The typical term for this is a "ragged array" I think.
